lookout/style/format/feature_extractor.py

```diff
@@ -796,15 +796,136 @@ def _compute_labels_mappings(self, vnodes: Iterable[VirtualNode]) -> None:
             else:
-                parent = (self._find_parent(j, file_vnodes, file_parents, closest_left_node_id)
-                          or uast)
-                vnode_parents[id(vn)] = parent
+                parent = (self._find_parent(annotation.stop, file, closest_left_node_id)
+                          or uast_annotation.uast)
+                file.add(TokenParentAnnotation(*annotation.span, parent))
+
+
+def _to_position(raw_lines_data, _lines_start_offset, offset):
+    line_num = numpy.argmax(_lines_start_offset > offset) - 1
+    col = offset - _lines_start_offset[line_num]
+    line = raw_lines_data[line_num]
+    if len(line) == col:
+        if line.splitlines()[0] != line:
+            # ends with newline
+            line_num += 1
+            col = 0
+    return Position(offset, line_num + 1, col + 1)
+
+
+def file_to_old_parse_file_format(file: AnnotationManager) -> Tuple[List["VirtualNode"],
+                                                                    Dict[int, bblfsh.Node]]:
+    """
+    Convert `AnnotationManager` instance to the deprecated output format of \
+    `FeatureExtractor._parse_file()`.
+
+    The function is created for backwards capability and should be removed after refactoring is \
+    finished.
+
+    :param file: file annotated with `UASTAnnotation`, `PathAnnotation` and `RawTokenAnnotation`. \
+                 It is expected to be the output of  `FeatureExtractor._parse_file()`.
+    :return: The old `FeatureExtractor._parse_file()` output format, that is \
+             Tuple with `VirtualNode`-s and `bbfsh.Node` id to parent mapping.
+    """
+    vnodes = []
+    path = file.get(PathAnnotation).path
+    raw_lines_data = file.sequence.splitlines(keepends=True)
+    line_lens = [0] + [len(d) for d in raw_lines_data]
+    line_lens[-1] += 1
+    _line_start_offsets = numpy.array(line_lens).cumsum()
+    for annotation in file.iter_by_type(RawTokenAnnotation):
+        vnode = VirtualNode(
+            file[annotation.span],
+            _to_position(raw_lines_data, _line_start_offsets, annotation.start),
+            _to_position(raw_lines_data, _line_start_offsets, annotation.stop),
+            is_accumulated_indentation=False,
+            path=path,
+            node=annotation.node,
+            y=None,
+        )
+        vnodes.append(vnode)
+    return vnodes, file.get(UASTAnnotation).parents
+
+
+def _file_to_vnodes_and_parents(file: AnnotationManager) -> Tuple[List["VirtualNode"],
+                                                                  Dict[int, bblfsh.Node]]:
+    """
+    Convert `AnnotationManager` instance to the deprecated format, which is a sequence of vnodes \
+    and vnodes parents mapping.
+
+    Used by `files_to_old_parse_vnodes_format` to generate old `_parse_vnodes` method output
+    format.
+
+    The function is created for backwards capability and should be removed after refactoring is \
+    finished.
```

Comment:
The docstring should explain how it is different from `file_to_old_parse_file_format`: the input and output types are exactly the same.